### ISSUE TITLE
torch.nn.functional.layer_norm returns nan for fp16 all 0 tensor

### ISSUE THREAD

**Issue by bbfrog** [pytorch/pytorch#41527]
## 🐛 Bug

When the input is a torch.float16 tensor and all values are 0, the torch.nn.functional.layer_norm function returns nan. It can be repro in pytorch 1.4.0 and pytorch 1.5.1 (haven't tried newer version), while pytorch 1.3.1 has no problem (return all 0 tensor).

## To Reproduce

The below sample code can repro: the last line prints a tensor with all nan values.
import torch

dim = 10
input =  torch.zeros(1, dim).float().cuda()
normalized_shape = tuple((dim,))
weight = torch.ones(dim).float().cuda()
bias = torch.zeros(dim).float().cuda()
eps=1e-12

output = torch.nn.functional.layer_norm(input, normalized_shape, weight, bias, eps)
print(output)

output = torch.nn.functional.layer_norm(input.half(), normalized_shape, weight.half(), bias.half(), eps)
print(output)

## Expected behavior

Current print results:
tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]], device='cuda:0')
tensor([[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]], device='cuda:0', dtype=torch.float16)

Expected to be
tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]], device='cuda:0')
tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]], device='cuda:0', dtype=torch.float16) 

## Environment

Collecting environment information...
PyTorch version: 1.5.1
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Enterprise Linux Server 7.2 (Paladin)
GCC version: (GCC) 5.4.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration:
GPU 0: Tesla V100-PCIE-32GB
GPU 1: Tesla V100-PCIE-32GB
GPU 2: Tesla V100-PCIE-32GB
GPU 3: Tesla V100-PCIE-32GB

Nvidia driver version: 440.33.01
cuDNN version: /home/common/cuda_toolkit/cudnn_7.4.2/cuda10.0/lib64/libcudnn.so.7.4.2

Versions of relevant libraries:
[pip] Could not collect
[conda] cudatoolkit               10.2.89              hfd86e86_1    anaconda
[conda] numpy                     1.18.1                   pypi_0    pypi
[conda] pytorch-memlab            0.1.0                    pypi_0    pypi
[conda] torch                     1.5.1                    pypi_0    pypi
[conda] torchvision               0.6.1                    pypi_0    pypi

## Additional context

cc @ezyang @gchanan @zou3519 @albanD @mruberry

**Comment by vadimkantorov** [comment]
Too small eps?

**Comment by bbfrog** [comment]
> Too small eps?

Thanks Vadimkantorov. You have a very good point. And yes, the eps is major reason: eps<=1e-8 can repro the problem, while eps>=1e-7 resolves the problem. It should align the float16 lower bound 5.96 e-8。

Increasing eps is a workaround for me. Thanks Vadimkantorov again very much.

@pytorch team, seems it is still a regression as pytorch 1.3.1 has no problem. Anyway the workaround is good for me.

**Comment by vadimkantorov** [comment]
I stumbled on this as well in https://github.com/pytorch/pytorch/issues/35666 :)

**Comment by zou3519** [comment]
Marking this tentatively high-pri for regression

**Comment by Laksh1997** [comment]
I've been using transformers recently with high learning rates and I get this issue (they have LayerNorms - when I switch to fp32 everything works fine)

**Comment by albanD** [comment]
Do I correctly understand that you force the eps value to a very small values (below the default one).
I think this is expected as the eps is here to control numerical stability and changing that value can change the behavior.

That solves your issue right?

**Comment by vadimkantorov** [comment]
I think it's inverse - forcing eps to a higher value instead.

One problem is that default is a fixed eps and not a polymorphic version at least retrieving eps from torch.finfo(..).tiny. So when some model is converted with some amp, this becomes a problem.

**Comment by albanD** [comment]
> I think it's inverse - forcing eps to a higher value instead.

But the default value for both the nn.Module and the functional is `1e-5`. That should be high enough no

> One problem is that default is a fixed eps 

Yes I do think this is an issue but we would need a more general solution.
But `layer_norm()` should not have this issue given the current default right?

**Comment by vadimkantorov** [comment]
@albanD you are right, for some reason the author originally used `eps=1e-12`, and default seems `1e-5`

**Comment by albanD** [comment]
Ok.
I'll close this as layer_norm should be fine here.
Forcing a very small epsilon (especially when using float16) is expected to lead to nan. You should not do that.

The general handling of the eps in the codebase though is much more complex and tricky because some are in python, some are in cpp, some can be controlled by the user or not.
But I think we would be happy to have an issue tracking these different cases and discussing what could be the way forward here.

**Comment by vadimkantorov** [comment]
about way forward discussion I had an earlier existing issue: https://github.com/pytorch/pytorch/issues/35666
